nemo_skills/dataset/mrcr/prepare.py

```diff
@@ -0,0 +1,129 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import subprocess
+import json
+from pathlib import Path
+import argparse
+from datasets import load_dataset
+from tqdm import tqdm
+import tempfile
+
+subprocess.run(["pip install tiktoken"], check=True, shell=True)
+import tiktoken
+"""
+Usage 
+# default. setup is all. 
+python prepare.py 
+
+# prepare subset needle2_128k. 
+python prepare.py --max_context_window 131072 --needles_subset 2 --setup needle2_128k
+python prepare.py --max_context_window 131072 --needles_subset 2 4 --setup needle2_needle_4_128k
+"""
+    
+    
+    
+def count_n_tokens(messages : list[dict]) -> int:
+    """
+    Follow the official way to count tokens in messages.
+    with tokenizer o200k_base
+    """
+    enc = tiktoken.get_encoding("o200k_base") 
+    return sum([len(enc.encode(m["content"])) for m in messages])
+
+
+def convert_messagesto_string(
+    messages : list[dict],
+    role_to_string : dict[str, str] = {"user": "user", "assistant": "assistant"},
+    ) -> str:
+    
+    return "\n".join([f"{role_to_string[m['role']]}: {m['content']}" for m in messages])
+
+
+def write_data_to_file(output_file, data, max_context_window, needles_subset, 
+                       convert_messagesto_string=False):
+    with open(output_file, "wt", encoding="utf-8") as fout:
+        for idx, entry in tqdm(enumerate(data), desc=f"Writing {output_file.name}"):
+            messages = json.loads(entry["prompt"])
+            
+            if entry['n_needles'] not in needles_subset:
+                print(f"Skipping {idx} because it has {entry['n_needles']} needle")
+                continue
+            
+            # find n_tokens
+            n_tokens = count_n_tokens(messages)
+            if max_context_window is not None:
+                if n_tokens > max_context_window:
+                    print(f"Skipping {idx} because it has {n_tokens} tokens")
+                    continue
+
+            # convert messages to string
+            if convert_messagesto_string:
+                question = convert_messagesto_string(messages, {"user": "User", "assistant": "Assistant"})
+            
+            else:
+                question = entry.pop('prompt')
+            entry['question'] = question
+            entry['expected_answer'] = entry.pop('answer')
+            entry['n_tokens'] = n_tokens
+            json.dump(entry, fout)
+            fout.write("\n")
+
+
+def get_mrcr_data(needles_subset, setup, max_context_window):
+        
+    dataset = load_dataset("openai/mrcr")['train']
+    data_dir = Path(__file__).absolute().parent
+    data_dir.mkdir(exist_ok=True)
+    
+    output_file = data_dir / f"{setup}.jsonl"
+    
+    with open(data_dir / "__init__.py", "w", encoding="utf-8") as init_file:
+        init_file.write(f"EVAL_SPLIT = '{setup}'\n")
```

Comment:
best not to override init dynamically here. Users can always provide `--split` argument to change this, so no need to change defaults

Reply:
Got it. Changed!